### PR TITLE
Eventhandler rework

### PR DESCRIPTION
**Pull request by StrongCoding**

### CODE REVIEW

File: incl/Cgi.hpp [hunk 1]
```diff
@@ -4,39 +4,62 @@
...
-	char **_enviromentVariables;
-	size_t _bodyLength;
-	std::string _clientIp;
+	State _state;
+	std::string _serverToCgiBuffer;
+	pid_t _childPid;
+
+	EventsData *_eventData;
 
 private:
```
Comment (wrottger) [inline incl/Cgi.hpp]: why again private? and why is private above public?

File: incl/Client.hpp [hunk 2]
```diff
@@ -34,22 +41,21 @@ class Client {
...
 
-	int getFd() const;
 	void updateTime();
 	bool isHeaderComplete() const;
-	void parseBuffer(const char *buffer);
 	std::time_t getLastModified() const;
-	HttpHeader *getHeaderObject() const;
 	void readFromClient();
 
 private:
```
Comment (wrottger) [inline incl/Client.hpp]: same thing here?

File: incl/Logger.hpp [hunk 3]
```diff
@@ -154,27 +154,25 @@ namespace Logging {
...
 	private:
-		static Logger* _instance;
+		static Logger _instance;
 		std::ofstream _logFile;
 		LogLevel _logLevel;
 		LogTarget _logTarget;
 		bool _timeStampInLog;
 		bool _logLevelInLog;
 
 	private:
```
Comment (wrottger) [inline incl/Logger.hpp]: WHYY?? @stevebalk

File: src/Client.cpp [hunk 4]
```diff
@@ -9,81 +9,80 @@ std::string createTestResponse() {
 	std::ostringstream oss;
 	oss << responseBody.size();
```
Comment (wrottger) [inline src/Client.cpp:10]: Here you could use the Utils

File: src/Client.cpp [hunk 5]
```diff
@@ -9,81 +9,80 @@ std::string createTestResponse() {
...
-	if (httpResponse != NULL) {
-		delete httpResponse;
+	LOG_DEBUG_WITH_TAG("Client destructor called", "Client");
+	if (_responseHttp != NULL) {
+		delete _responseHttp;
+	}
+	if (_responseCgi != NULL) {
+		delete _responseCgi;
 	}
-	delete _headerObject;
```
Comment (wrottger) [inline src/Client.cpp:37]: You don't need to NULL check before deleting

File: src/ConfigUtils.cpp [hunk 6]
```diff
@@ -175,25 +176,97 @@ std::string Config::getErrorPage(int code, const std::string& route, const std::
...
 }
 
-bool Config::isCGIAllowed(const std::string& route, const std::string& host)
+std::string Config::getCgiScriptPath(const HttpHeader &header)
 {
     Config* config = getInstance();
     if (config == NULL) //prevent segfault
         throw std::runtime_error("Cannot use isCGIAllowed without a valid Config instance.");
-    std::string path;
+    std::string path = "";
```
Comment (wrottger) [inline src/ConfigUtils.cpp]: the default constructor already sets it to an empty string

File: src/ConfigUtils.cpp [hunk 7]
```diff
@@ -175,25 +176,97 @@ std::string Config::getErrorPage(int code, const std::string& route, const std::
...
+    int file = 0;
```
Comment (wrottger) [inline src/ConfigUtils.cpp]: could this be a bool? as an int I mistook it for a file descriptor.

File: src/EventHandler.cpp [hunk 8]
```diff
@@ -98,56 +112,61 @@ EventsData *EventHandler::createNewEvent(int fd, EventType type, Client *client)
...
+// Unregister event from epoll
 void EventHandler::unregisterEvent(EventsData *eventData) {
 	if (epoll_ctl(_epollFd, EPOLL_CTL_DEL, eventData->fd, NULL) == -1) {
 		perror("epoll_ctl");
 		std::cerr << "Fd: " << eventData->fd << std::endl; // TODO: DELETE
 		LOG_ERROR("UnregisterEvent: epoll DEL failed.");
 		perror("epoll_ctl");
 	}
-	for (std::list<EventsData *>::iterator it = _eventDataList.begin(); it != _eventDataList.end(); it++) {
+	for (std::list<EventsData *>::iterator it = _eventDataList.begin(); it != _eventDataList.end();) {
```
Comment (wrottger) [inline src/EventHandler.cpp]: Could you write this using std::find_if() ?

File: src/Client.cpp (hunk 5 above)
Comment (StrongCoding) [reply src/Client.cpp:37]: if you set them to NULL after deleting you dont need to check again

File: incl/Logger.hpp (hunk 3 above)
Comment (stevebalk) [reply incl/Logger.hpp]: Changed to Meyers' Singleton pattern, also deleted unused _instance variable 2 minutes ago.
See also the change in Logger.cpp -> https://github.com/wrottger/webserv/blob/d6b0a879b311c2bde2d898c7026160662b22d32c/src/Logger.cpp#L18C1-L22C2

File: src/Client.cpp (hunk 5 above)
Comment (stevebalk) [reply src/Client.cpp:37]: Removed redundant NULL check

**Review by stevebalk — APPROVED** [review]

**Pull request merged**